### PR TITLE
Laboratory work #2, Anna Vorontsova - 22FPL2

### PR DESCRIPTION
**Pull request by vorontsann**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
👀

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -14,7 +14,18 @@ def prepare_word(
     :param end_of_word: a token that signifies the end of word
     :return: preprocessed word
     """
-
+    if (not isinstance(raw_word, str) or
+            not (isinstance(start_of_word, str) or start_of_word is None) or
+            not (isinstance(end_of_word, str) or end_of_word is None)):
+        return None
+
+    if not start_of_word and not end_of_word:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:22]: too dificult

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -26,6 +37,21 @@ def collect_frequencies(
...
     """
+    if (not isinstance(text, str) or
+            not (isinstance(start_of_word, str) or start_of_word is None) or
+            not isinstance(end_of_word, str)):
+        return None
+
+    dict_of_freq = {}
+    words = text.split()
+    for word in words:
+        prepr_word = prepare_word(word, None, '</s>')
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:48]: are you sure? Why do not you use arguments?

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -26,6 +37,21 @@ def collect_frequencies(
...
+        if prepr_word is None:
+            return None
+        if prepr_word not in dict_of_freq:
+            dict_of_freq[prepr_word] = words.count(word)
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:52]: you add the same many times

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (vorontsann) [reply lab_2_tokenize_by_bpe/main.py:24]: fixed?

File: lab_2_tokenize_by_bpe/main.py (hunk 2 above)
Comment (vorontsann) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 3 above)
Comment (vorontsann) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (artyomtugaryov) [reply lab_2_tokenize_by_bpe/main.py:24]: do not think so

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -36,6 +63,22 @@ def count_tokens_pairs(
...
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+
+    result_dic = {}
+    for pair in word_frequencies.items():
+        word = pair[0]
+        count = pair[1]
+        for i in range(len(word) - 1):
+            token1 = word[i]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:74]: what about slices?

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -47,6 +90,23 @@ def merge_tokens(
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if (not isinstance(word_frequencies, dict) or
+            not isinstance(pair, tuple)):
+        return None
+
+    new_word_freq = {}
+    for pairs in word_frequencies.items():
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:98]: create `word`, `count` right here

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 4 above)
Comment (vorontsann) [reply lab_2_tokenize_by_bpe/main.py:74]: sorry i don't understand what slices do you mean...

File: lab_2_tokenize_by_bpe/main.py (hunk 5 above)
Comment (vorontsann) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 4 above)
Comment (artyomtugaryov) [reply lab_2_tokenize_by_bpe/main.py:72]: `pair = word[i:i+2]`

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -47,6 +88,26 @@ def merge_tokens(
...
+    if (not isinstance(word_frequencies, dict) or
+            not isinstance(pair, tuple)):
+        return None
+
+    new_word_freq = word_frequencies.copy()
+    for word in word_frequencies:
+        new_word = []
+        if pair[0] in word and pair[1] in word:
+            for i in range(len(word) - 1):
+                if word[i] == pair[1] and word[i - 1] == pair[0]:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:100]: why do you check by elements? Looks like you can compare full tuples

File: lab_2_tokenize_by_bpe/main.py [hunk 7]
```diff
@@ -58,6 +119,32 @@ def train(
...
+            not isinstance(num_merges, int)):
+        return None
+
+    dict_pairs = count_tokens_pairs(word_frequencies)
+    if dict_pairs is None:
+        return None
+    num_merges = min(num_merges, len(dict_pairs))
+
+    for iteration in range(num_merges):
+        if dict_pairs != {}:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:132]: bad check

File: lab_2_tokenize_by_bpe/start.py [hunk 8]
```diff
@@ -12,8 +14,11 @@ def main() -> None:
     with open(assets_path / 'text.txt', 'r', encoding='utf-8') as text_file:
         text = text_file.read()
 
-    result = None
-    assert result, "Encoding is not working"
+    word_freq = main_file.collect_frequencies(text, None, '</s>')
+    print(main_file.train(word_freq, 100))
+
+    # result = main_file.train(word_freq, 100)
+    # assert result, "Encoding is not working"
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/start.py:21]: NOOOO!

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 6 above)
Comment (vorontsann) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 7 above)
Comment (vorontsann) [reply lab_2_tokenize_by_bpe/main.py]: fixed

**Pull request closed without merging**